lib/juttle-service-client.js

```diff
@@ -0,0 +1,347 @@
+                                                  modules: err_obj.info.bundle.modules,
+                                                  filename: filename}));
+    } else {
+        write_error(sprintf('ERROR %s', err.message));
+    }
+}
+
+function read_inputs(opts) {
+
+    let inputs = {};
+    if (opts.input !== undefined) {
+        if (! _.isArray(opts.input)) {
+            opts.input = [opts.input];
+        }
+        _.each(opts.input, function(x) {
+            let L = x.split('=');
+            if (L.length !== 2) {
+                write_error(sprintf('error: invalid input %s', x));
+                usage();
+            }
+            let num = +L[1];
+            inputs[L[0]] = isNaN(num) ? L[1] : num;
+        });
+    }
+
+    return inputs;
+}
+
+let cmdargs = {
+    'path': {
+        usage: '--path <path-to-juttle-file>',
+        desc: 'Path to file relative to configured root directory.'
+    },
+    'wait': {
+        usage: '--wait',
+        desc: 'If true, wait for program to finish, otherwise run in background'
+    },
+    'input': {
+        usage: '--input name=val',
+        desc: 'One or more input values.'
+    },
+    'job': {
+        usage: '--job <job-id>',
+        desc: 'Job id.'
+    },
+    'observer': {
+        usage: '--observer <observer-id>',
+        desc: 'Observer id.'
+    },
+    'server': {
+        usage: '--juttle-service <hostname:port>',
+        desc: 'Hostname/port of server'
+    },
+    help: {
+        usage: '--help',
+        desc: 'Print this help and exit'
+    }
+}
+
+let commands = {
+    subscribe: {
+        usage: '({job} | {observer})',
+        perform: subscribe
+    },
+    list_jobs: {
+        usage: '[{job}]',
+        perform: list_jobs
+    },
+    list_observers: {
+        usage: '',
+        perform: list_observers
+    },
+    run: {
+        usage: '{path} [{wait}] [{observer}]',
+        perform: run
+    },
+    delete: {
+        usage: '{job}',
+        perform: delete_job
+    },
+    get_inputs: {
+        usage: '{path} {input} [{input} ...]',
+        perform: get_inputs
+    }
+};
+
+function expand(string) {
+    return string.replace(/\{([^}]*)\}/g, function(match, key) {
+        return cmdargs[key].usage;
+    });
+}
+
+function init(options) {
+    write_output = options.write_output;
+    write_error = options.write_error;
+    client_exit = options.client_exit;
+    commands = _.extend(commands, options.commands);
+    cmdargs = _.extend(cmdargs, options.cmdargs);
+}
+
+function usage() {
+
+    // Given the command descriptions, create the list of commands
+    // used by each option.
+    _.each(cmdargs, function (arginfo, arg) {
+        arginfo.commands = [];
+        _.each(commands, function(info, cmd) {
+            if (info.usage.indexOf('{' + arg + '}') > -1) {
+                arginfo.commands.push(cmd);
+            }
+        });
+    });
+
+    // Expand each command usage
```

Comment:
To me that part of the usage isn't all that useful, but in the interest of
getting this done I can let it go.

-m

On Tue, Feb 9, 2016 at 6:03 PM, Mark Stemm <notifications@github.com> wrote:

> In lib/juttle-service-client.js
> <https://github.com/juttle/juttle-service/pull/22#discussion_r52406440>:
>
> > +}
> > +
> > +function usage() {
> > +
> > +    // Given the command descriptions, create the list of commands
> > +    // used by each option.
> > +    _.each(cmdargs, function (arginfo, arg) {
> > +        arginfo.commands = [];
> > +        _.each(commands, function(info, cmd) {
> > +            if (info.usage.indexOf('{' + arg + '}') > -1) {
> > +                arginfo.commands.push(cmd);
> > +            }
> > +        });
> > +    });
> > +
> > +    // Expand each command usage
>
> One reason you can't just blindly copy is that the part of the usage:
>
> --job <job-id> Job id.
> used by: subscribe,list_jobs,delete
> where each command line argument names the commands it's used for.
>
> If juttle-engine adds another command that also uses --job as a command
> line argument, I think you have to do some merging.
>
> But if you mean just the expansion, I can take that out.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/juttle/juttle-service/pull/22/files#r52406440>.
>



-- 
Michael Demmer
Jut - CTO / VP Engineering
demmer@jut.io